### PR TITLE
CRL: Ignore expired certificates

### PR DESCRIPTION
**Pull request by reinkrul**
Fixes https://github.com/nuts-foundation/nuts-node/issues/1761 and https://github.com/nuts-foundation/nuts-node/issues/1762

### CODE REVIEW

**Comment by dmeijboom** [comment]
I see that a review was requested from me so I shall respond, whether it was by accident or not :P

In the `IsSynced` method for loop is changed to this:

```go
endpoints := v.parseCRLEndpoints()

for endpoint, issuer := range endpoints {
```

I'm not sure if that's correct as `parseCRLEndpoints` returns a map of an CRL endpoint with the certificate as it's value. The certificate doesn't have to be the issuer of the CRL, it could be any certificate in the chain that has the CRL endpoint configured.

> // Certificate has expired, no need to update CRL since it can't sign any new CRLs.

The CRL is signed by the CA or an intermediate CA. If you grab any certificate and download it's CRL it should be signed by the issuer of that certificate. That's also what has been implemented in the `verifyCRL` function:

```go
	certificate, ok := v.certificates[issuerName]
	if !ok {
		return errors.New("CRL signature could not be validated against known certificates")
	}

	return certificate.CheckCRLSignature(crl)
```

So I doubt this is correct as the certificate doesn't have to be the certificate that signed the CRL.

File: crl/validator.go [hunk 1]
```diff
@@ -143,7 +148,7 @@ func (v *validator) verifyCRL(crl *pkix.CertificateList) error {
 	return certificate.CheckCRLSignature(crl)
```
Comment (dmeijboom) [inline crl/validator.go:148]: Not sure if Nuts is already using Go 1.19 but `Certificate.CheckCRLSignature` is deprecated:

> Deprecated: Use RevocationList.CheckSignatureFrom instead.

Comment (reinkrul) [reply crl/validator.go:148]: Yeah, I noticed. More work :'-)

**Comment by woutslakhorst** [comment]
+backport to V5

**Comment by woutslakhorst** [comment]
> In the IsSynced method for loop is changed to this:
>
> endpoints := v.parseCRLEndpoints()
> 
> for endpoint, issuer := range endpoints {
> I'm not sure if that's correct as parseCRLEndpoints returns a map of an CRL endpoint with the certificate as it's value. The certificate doesn't have to be the issuer of the CRL, it could be any certificate in the chain that has the CRL endpoint configured.

It might just be the `issuer` term which refers to the CA and not the issuer of the CRL?

**Comment by dmeijboom** [comment]
> > In the IsSynced method for loop is changed to this:
> > endpoints := v.parseCRLEndpoints()
> > for endpoint, issuer := range endpoints {
> > I'm not sure if that's correct as parseCRLEndpoints returns a map of an CRL endpoint with the certificate as it's value. The certificate doesn't have to be the issuer of the CRL, it could be any certificate in the chain that has the CRL endpoint configured.
> 
> It might just be the `issuer` term which refers to the CA and not the issuer of the CRL?

Iirc the certificates stored in the CRL validator could be any certificate. Whether that's a CA, intermediate CA or a normal certificate. There is (as far as I'm aware) no direct relationship to the issuer. Otherwise the code in `verifyCRL` would have been really straightforward (it wouldn't have to look up the issuer).

**Comment by reinkrul** [comment]
The signer of the CRL can be, indeed, any (trusted) CA:

>  A CRL is issued by a CRL issuer, which is typically the CA which also issued the corresponding certificates, but could alternatively be some other trusted authority.

(https://en.wikipedia.org/wiki/Certificate_revocation_list)

Fixing this actually made the code a bit simpler and more correct; now it takes into account multiple CA certificates using the same CRL endpoint. If there are no (more) active CA certificates using the CRL endpoint (in other words, all expired), it skips the CRL endpoint.

**Review by woutslakhorst — APPROVED** [review]

**Pull request merged**